### PR TITLE
Rename resources. closes #1080

### LINKED ISSUE #1080: Finalize naming of resources in servicecatalog API group
This issue is to finalize the names of API resources before moving the service-catalog API to beta.

### Requirements:

- We should not have any name collisions with core resources
- When naming resources, we should run the thought experiment of "would we like it if everyone followed our naming patterns?" - we are likely to establish precedent for how other extension APIs name their resources
- We should only do a single rename if at all possible - iterating on names over multiple releases isn't good for users

### Suggested process:

1. Find a consensus within the group
2. Ensure that our names pass API review
3. Make a proposal document for v1beta1 API that we can link users to
4. Communicate proactively about the breaking API changes and give users a specific release date when the changes will be released
5. Make the code change PR

#### Note

We have previously had a number of issues for different objectives or pieces of the rename, and a discussion wherein there was a miscommunication about what we were 
discussing.  I've closed the previous issues we've had for this topic so that we can get a fresh start.  Previous issues were:

- #837
- #1008
- #1031
- #1033
- #1060
- #1054

### Proposed resources

Based on our prior discussions in the group, the set of resources that will be the 'nucleus' of the catalog will be as follows.  Note: names used here are purely to identify the resources and not proposed names:

- Broker
- ServiceClass
- ServicePlan
- Instance
- Binding (we have discussed calling this InstanceCredential, but no one seemed to be especially happy with the name)

Based on our current knowledge, the above should be a complete accounting of the resources in the v1beta1 API, but there may be additional resources pending the outcome of #802.

Note: The binding resource should _only_ represent the OSB facet of binding, see #994.

#### Extension resources

There are a number of different extension resources that we know there is interest in building in the community:

- A resource that manages a PodPreset based on the lifecycle of our binding resource
- A resource that does a `VCAP_SERVICES`-style injection
- A resource that creates Istio rules
- A 'flex'-style binding that provides an easy on-ramp for people to extend the catalog

These should be created initially in (a) **different API group(s)** from the nucleus of the service-catalog.  It might be convenient for some of these resources with wide appeal to be created in this repository, and more appropriate for others to be created outside of this repo.

### Proposed resource names

I propose the following resource names:

- `ServiceBroker`
- `ServiceClass`
- `ServicePlan`
- `ServiceInstance`
- `ServiceInstanceCredential`

#### Previous discussion: naming of binding resource

See #1033

#### Previous discussion: separating service classes and plans

See #454

#### Previous discussion: `ServiceCatalog` prefix

We had previously discussed prefixing every resource with `ServiceCatalog`.  I think there are a number of cons to this:

- Extremely lengthy names to type in a CLI
- No precedent for encoding name of API group in every resource name in that group
- Not a good precedent to establish: if all other extension APIs follow this pattern, the entire ecosystem will be filled with extremely lengthy names
- Yields difficult to understand resource names

Since this is necessarily an issue to resolve and implement before beta, I'm adding it to the 0.1.0 milestone.

### PR DESCRIPTION
**Pull request by vaikas-google**
Rename all the resources to have the consistent prefix of Service:
instance -> serviceinstance
binding -> serviceinstancecredential
broker -> servicebroker

Closes #1080

### CODE REVIEW

**Comment by bmelville** [comment]
image

**Comment by pmorie** [comment]
\m/ impressive

**Comment by MHBauer** [comment]
Nothing serious follows. Good Luck!

It already needs a rebase and rebasing 1106 is going to be :-( . Can't we just edit some json templates? I feel like we need a layer in between to map our external to internal. JUST ONE MORE LAYER OF INDIRECTION GUYS!

**Comment by pmorie** [comment]
> It already needs a rebase and rebasing 1106 is going to be :-( 

Yeah :( On the plus side, kube will turn you into a rebase champ.  I'm happy to help with any rebase jobs as needed.

File: cmd/controller-manager/app/options/options.go [hunk 1]
```diff
@@ -88,7 +88,7 @@ func (s *ControllerManagerServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.ServiceCatalogKubeconfigPath, "service-catalog-kubeconfig", "", "Path to service-catalog kubeconfig")
 	fs.BoolVar(&s.ServiceCatalogInsecureSkipVerify, "service-catalog-insecure-skip-verify", s.ServiceCatalogInsecureSkipVerify, "Skip verification of the TLS certificate for the service-catalog API server")
 	fs.DurationVar(&s.ResyncInterval, "resync-interval", s.ResyncInterval, "The interval on which the controller will resync its informers")
-	fs.DurationVar(&s.BrokerRelistInterval, "broker-relist-interval", s.BrokerRelistInterval, "The interval on which a broker's catalog is relisted after the broker becomes ready")
+	fs.DurationVar(&s.ServiceBrokerRelistInterval, "broker-relist-interval", s.ServiceBrokerRelistInterval, "The interval on which a broker's catalog is relisted after the broker becomes ready")
```
Comment (pmorie) [inline cmd/controller-manager/app/options/options.go]: We might want to keep the name of this flag as is, or we're going to make a nice little mess for ourselves in the helm chart.

Comment (pmorie) [inline cmd/controller-manager/app/options/options.go]: We might want to keep the name of this flag as is, or we're going to make a nice little mess for ourselves in the helm chart.

File: docs/design.md [hunk 2]
```diff
@@ -33,16 +33,16 @@ them.
...
-  reference and use a particular Service Instance.
-- **Broker**, or *Service Broker* : a entity, available via a web endpoint,
+  reference and use a particular Service ServiceInstance.
+- **ServiceBroker**, or *Service Broker* : a entity, available via a web endpoint,
   that manages a set of one or more Services.
 - **Credentials** : Information needed by an Application to talk with a
-  Service Instance.
-- **Instance**, or *Service Instance* : Each independent use of a Service
-  Class is called a Service Instance.
+  Service ServiceInstance.
```
Comment (pmorie) [inline docs/design.md]: Service ServiceInstance

File: docs/design.md [hunk 3]
```diff
@@ -33,16 +33,16 @@ them.
...
   than Service Catalog does, so to avoid confusion the term *Application*
   will refer to the Kubernetes deployment artifact that will use a Service
-  Instance.
-- **Binding**, or *Service Binding* : a link between a Service Instance
+  ServiceInstance.
+- **ServiceInstanceCredential**, or *Service ServiceInstanceCredential* : a link between a Service ServiceInstance
   and an Application. It expresses the intent for an Application to
-  reference and use a particular Service Instance.
-- **Broker**, or *Service Broker* : a entity, available via a web endpoint,
+  reference and use a particular Service ServiceInstance.
```
Comment (pmorie) [inline docs/design.md]: Service ServiceInstance

File: docs/design.md [hunk 4]
```diff
@@ -33,16 +33,16 @@ them.
 - **Application** : Kubernetes uses the term "service" in a different way
   than Service Catalog does, so to avoid confusion the term *Application*
   will refer to the Kubernetes deployment artifact that will use a Service
-  Instance.
-- **Binding**, or *Service Binding* : a link between a Service Instance
+  ServiceInstance.
+- **ServiceInstanceCredential**, or *Service ServiceInstanceCredential* : a link between a Service ServiceInstance
```
Comment (pmorie) [inline docs/design.md]: sed left some weird stuff here

File: docs/design.md [hunk 5]
```diff
@@ -159,42 +159,42 @@ Users can then query for the list of available Services:
 
     kubectl get services
 
-### Creating a Service Instance
+### Creating a Service ServiceInstance
```
Comment (pmorie) [inline docs/design.md]: Service ServiceInstance

File: docs/design.md (hunk 4 above)
Comment (vaikas-google) [inline docs/design.md]: Yup, I think I got them all fixed now.

File: cmd/controller-manager/app/options/options.go (hunk 1 above)
Comment (pmorie) [inline cmd/controller-manager/app/options/options.go]: oops, my bad.  This should be totally fine as-is

**Comment by pmorie** [comment]
LGTM, :100:

File: contrib/jenkins/test_walkthrough.sh [hunk 6]
```diff
@@ -179,15 +179,15 @@ kubectl --context=service-catalog create -f "${ROOT}/contrib/examples/walkthroug
   || error_exit 'Error when creating ups-binding.'
 
 wait_for_expected_output -e 'InjectedBindResult' \
-  kubectl --context=service-catalog get bindings -n test-ns ups-binding -o yaml \
+  kubectl --context=service-catalog get serviceinstancecredentials -n test-ns ups-binding -o yaml \
```
Comment (kibbles-n-bytes) [inline contrib/jenkins/test_walkthrough.sh]: We'll also want to change "ups-binding" to "ups-instance-credential". Fine in a follow-up.

File: docs/walkthrough.md [hunk 7]
```diff
@@ -111,38 +111,38 @@ check out the "Cleaning Up" section below. Follow those instructions before you
...
+- `ServiceInstance`: A provisioned instance of a `ServiceClass`. These are created
 by cluster users who wish to make a new concrete _instance_ of some _type_ of
 managed service to make that available for use by one or more in-cluster
-applications. When a new `Instance` resource is created, the service catalog
+applications. When a new `ServiceInstance` resource is created, the service catalog
 controller will connect to the appropriate broker server and instruct it to
 provision the service instance.
-- `Binding`: A "binding" to an `Instance`. These are created by cluster users
-who wish for their applications to make use of a service `Instance`. Upon
+- `ServiceInstanceCredential`: A "binding" to an `ServiceInstance`. These are created by cluster users
```
Comment (kibbles-n-bytes) [inline docs/walkthrough.md]: ```A "binding" to an `ServiceInstance` ``` -> ```Access credentials to a `ServiceInstance` ``` . Fine in a follow-up, for the sake of getting this in soon.

File: docs/walkthrough.md [hunk 8]
```diff
@@ -111,38 +111,38 @@ check out the "Cleaning Up" section below. Follow those instructions before you
...
+who wish for their applications to make use of a service `ServiceInstance`. Upon
 creation, the service catalog controller will create a Kubernetes `Secret`
 containing connection details and credentials for the service instance. Such
 `Secret`s can be mounted into pods as usual.
 
 These concepts and resources are the building blocks of the service catalog.
 
-## Step 3 - Installing the UPS Broker
+## Step 3 - Installing the UPS ServiceBroker
```
Comment (kibbles-n-bytes) [inline docs/walkthrough.md]: As a meta comment, we should later decide what exactly we want our messaging should be. I think there's a distinction between a "service broker" (the overall concept) and "ServiceBroker" (the API server resource). On top of that, I also think it's okay to have discussions about "brokers" (minus the "service" aspect) if the context is clear.

Definitely not in the scope of this PR, but I think we should later go through our docs with a fine-toothed comb with those sorts of things in mind.

**Comment by kibbles-n-bytes** [comment]
LGTM. RIP every outstanding PR (including mine :rofl: )

**Comment by kibbles-n-bytes** [comment]
@vaikas-google I just noticed now; some other things we'll want to change are:

- filenames under `pkg/controller/`
- queue names in `pkg/controller/controller.go`
- lister names in `pkg/controller/controller.go`
- informer function names (add/update/delete) (scattered throughout `pkg/controller`)

**Comment by vaikas-google** [comment]
Yes, I wanted to split those into smaller changes to get the major changes
in without crazy remerge issues.
I'll tackle those, one is already out. Others will follow.

**Pull request merged**